### DISCUSSION TITLE
Introducing Conditionals in PlayBooks

### DISCUSSION THREAD

**Discussion by sidPhoenix17** [DrDroidLab/PlayBooks#221] [General]
Hello everyone, we are introducing capability to setup conditional rules within PlayBooks. We are currently working on the design of the functionality and if any of you have any use-cases or comments, please do share!

<img width="809" alt="conditionals" src="https://github.com/DrDroidLab/PlayBooks/assets/16644933/d79aca26-bdfd-4675-9d58-02c043aa9889">

As part of conditional playbooks, a user will be able to to setup a step output based execution path of the playbook. This will help anyone to execute the playbooks step by step guided by configuration pre-created by the creator of the playbook.

While configuring the playbook steps, the user can do the following on top of the output:

- Select a function that will be applied to the step output. These functions could be:
    - For time series output, it will be an aggregation function on top of the time series to create the LHS of the expression. These can be:
        - Average
        - Max
        - Min
        - Latest
    - For logs & DB queries:
        - Count of results
        - Query result Value
        - Query result List of values
- User can then use the output of the function in a conditional expression. It can have the following operations:
    - For numerical LHS: 
        -- Equals (=) with RHS
        -- Comparison (<, <=, >, >=, between) with RHS
        -- Increased / Decreased (Check against previous window without RHS)
    - For string LHS:
        -- Equals (=) with RHS
        -- Regex / Substring Match with RHS
    - For List LHS:
        -- Contains the RHS
- Based on the expression, user can configure next step to execute. The next step will be a child to the original step and can’t be a child of any other step.
- Apart from setting up a conditional expression, user can also setup a function to just export the output into a variable for being used in other steps of the playbook executed subsequently.
- A playbook with any step having a conditional expression will have ‘Run All’ button disabled.
- Once someone lands on this playbook page, they click on ‘Start’ button and with each execution they are shown the output and what next step must be executed as per the conditional path.
- User can still choose to execute any step they want.
- All outputs, expression results and clicks will be recorded in the playbook timeline that can be accessed at any time for seeing investigation done so far or for review later. This timeline will be exportable into any RCA writing tool of your choice.

**Comment by ShubhamBhardwaj1971** [comment]
For string outputs, I will prefer to have Regex and In List options too for doing comparison.

**Reply by dimittal** [reply]
Great suggestions @ShubhamBhardwaj1971. Have incorporated Regex matching & In List operations in the first version.

**Reply by ShubhamBhardwaj1971** [reply]
Thanks Sir!
